pkg/controllers/externaloidc/externaloidc_controller.go

```diff
@@ -0,0 +1,363 @@
+}
+
+func (c *externalOIDCController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	if c.featureGateAccessor == nil || !c.featureGateAccessor.AreInitialFeatureGatesObserved() {
+		return nil
+	}
+
+	featureGates, err := c.featureGateAccessor.CurrentFeatureGates()
+	if err != nil {
+		return fmt.Errorf("could not get current feature gates: %v", err)
+	}
+
+	if !featureGates.Enabled(features.FeatureGateExternalOIDC) {
+		return nil
+	}
+
+	auth, err := c.authLister.Get("cluster")
+	if err != nil {
+		return fmt.Errorf("could not get authentication/cluster: %v", err)
+	}
+
+	if auth.Spec.Type != configv1.AuthenticationTypeOIDC {
+		// auth type is "IntegratedOAuth", "" or "None"; delete structured auth configmap if it exists
+		if err := c.configMaps.ConfigMaps(targetConfigMapNamespace).Delete(ctx, targetAuthConfigCMName, metav1.DeleteOptions{}); err == nil {
+			syncCtx.Recorder().Eventf(c.eventName, "Removed auth configmap %s/%s", targetConfigMapNamespace, targetAuthConfigCMName)
+
+		} else if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("could not delete existing configmap %s/%s: %v", targetConfigMapNamespace, targetAuthConfigCMName, err)
+		}
+
+		return nil
+	}
+
+	authConfig, err := c.generateAuthConfig(*auth)
+	if err != nil {
+		return err
+	}
+
+	authConfigJSON, err := json.Marshal(authConfig)
+	if err != nil {
+		return fmt.Errorf("could not marshal auth config into JSON: %v", err)
+	}
+
+	existingCM, err := c.configMapLister.ConfigMaps(targetConfigMapNamespace).Get(targetAuthConfigCMName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("could not retrieve auth configmap %s/%s to check data before sync: %v", targetConfigMapNamespace, targetAuthConfigCMName, err)
+	}
+
+	if existingCM != nil && existingCM.Data[authConfigDataKey] == string(authConfigJSON) {
+		return nil
+	}
+
+	errList := validateAuthenticationConfiguration(*authConfig)
+	if len(errList) > 0 {
+		return fmt.Errorf("auth config validation failed: %v", errList)
+	}
+
+	cm := corev1ac.ConfigMap(targetAuthConfigCMName, targetConfigMapNamespace).WithData(map[string]string{authConfigDataKey: string(authConfigJSON)})
+	if _, err := c.configMaps.ConfigMaps(targetConfigMapNamespace).Apply(ctx, cm, metav1.ApplyOptions{FieldManager: c.name}); err != nil {
+		return fmt.Errorf("could not apply changes to auth configmap %s/%s: %v", targetConfigMapNamespace, targetAuthConfigCMName, err)
+	}
+
+	syncCtx.Recorder().Eventf(c.eventName, "Synced auth configmap %s/%s", targetConfigMapNamespace, targetAuthConfigCMName)
+
+	return nil
+}
+
+// generateAuthConfig creates a structured JWT AuthenticationConfiguration for OIDC
+// from the configuration found in the authentication/cluster resource
+func (c *externalOIDCController) generateAuthConfig(auth configv1.Authentication) (*apiserverv1beta1.AuthenticationConfiguration, error) {
+	authConfig := apiserverv1beta1.AuthenticationConfiguration{
+		TypeMeta: metav1.TypeMeta{
```

Comment:
So, by side-stepping api-machinery and using `json.Marshal` we need to set it by hand. Do we do that in other places as well or do we leverage the serializer?

Reply:
`apiserverv1beta1.AuthenticationConfiguration` won't go via an API call, but instead it will be serialized into a json file and read back by the `kube-apiserver`, so for the moment there's no apimachinery to handle this AFAIK.

Reply:
It has a [register.go](https://github.com/openshift/cluster-authentication-operator/blob/master/vendor/k8s.io/apiserver/pkg/apis/apiserver/v1beta1/register.go#L55). We would need to create a runtime.Scheme and add the object to the scheme, to then be able to use the codec for serialization and deserialization? I am not sure, I am not an expert in api machinery topics and can't tell if it is "the better way" or if it adds too much complexity.

Reply:
Nice, wasn't very familiar with schemes, TIL :+1: Pushing a fix.

Reply:
Now, that we use the codec, we can drop the TypeMeta, right?